worker/rhpv2.go

```diff
@@ -331,83 +416,105 @@ func (s *Session) Read(ctx context.Context, w io.Writer, sections []rhpv2.RPCRea
+			// through a Merkle proof verifier before verifying the AEAD tag.
+			// Security therefore depends on the caller of Read discarding any data
+			// written to w in the event that verification fails.
+			msgReader, err := t.RawResponse(4096 + uint64(sec.Length))
+			if err != nil {
+				return wrapResponseErr(err, "couldn't read sector data", "host rejected Read request")
 			}
-		}
-		// verify the message tag and the Merkle proof
-		if err := msgReader.VerifyTag(); err != nil {
-			return err
-		}
-		if !rpv.Verify(proof, sec.MerkleRoot) {
-			return ErrInvalidMerkleProof
-		}
-		// if the host sent a signature, exit the loop; they won't be sending
-		// any more data
-		if hostSig != nil {
+			// Read the signature, which may or may not be present.
+			lenbuf := make([]byte, 8)
+			if _, err := io.ReadFull(msgReader, lenbuf); err != nil {
+				return fmt.Errorf("couldn't read signature len: %w", err)
+			}
+			if n := binary.LittleEndian.Uint64(lenbuf); n > 0 {
+				hostSig = new(types.Signature)
+				if _, err := io.ReadFull(msgReader, hostSig[:]); err != nil {
+					return fmt.Errorf("couldn't read signature: %w", err)
+				}
+			}
+			// stream the sector data into w and the proof verifier
+			if _, err := io.ReadFull(msgReader, lenbuf); err != nil {
+				return fmt.Errorf("couldn't read data len: %w", err)
+			} else if binary.LittleEndian.Uint64(lenbuf) != uint64(sec.Length) {
+				return errors.New("host sent wrong amount of sector data")
+			}
+			proofStart := sec.Offset / rhpv2.LeafSize
+			proofEnd := proofStart + sec.Length/rhpv2.LeafSize
+			rpv := rhpv2.NewRangeProofVerifier(proofStart, proofEnd)
+			tee := io.TeeReader(io.LimitReader(msgReader, int64(sec.Length)), &segWriter{w: w})
+			// the proof verifier Reads one segment at a time, so bufio is crucial
+			// for performance here
+			if _, err := rpv.ReadFrom(bufio.NewReaderSize(tee, 1<<16)); err != nil {
+				return fmt.Errorf("couldn't stream sector data: %w", err)
+			}
+			// read the Merkle proof
+			if _, err := io.ReadFull(msgReader, lenbuf); err != nil {
+				return fmt.Errorf("couldn't read proof len: %w", err)
+			}
+			if binary.LittleEndian.Uint64(lenbuf) != uint64(rhpv2.RangeProofSize(rhpv2.LeavesPerSector, proofStart, proofEnd)) {
+				return errors.New("invalid proof size")
+			}
+			proof := make([]types.Hash256, binary.LittleEndian.Uint64(lenbuf))
+			for i := range proof {
+				if _, err := io.ReadFull(msgReader, proof[i][:]); err != nil {
+					return fmt.Errorf("couldn't read Merkle proof: %w", err)
+				}
+			}
+			// verify the message tag and the Merkle proof
+			if err := msgReader.VerifyTag(); err != nil {
+				return err
+			}
+			if !rpv.Verify(proof, sec.MerkleRoot) {
+				return ErrInvalidMerkleProof
+			}
+			// if the host sent a signature, exit the loop; they won't be sending
+			// any more data
+			if hostSig != nil {
+				return errNoMoreData
+			}
+			return nil
+		}(); err == errNoMoreData {
```

Comment:
this isn't really an error. You can declare a `var noMoreData bool` above the closure, then `noMoreData = hostSig != nil` inside it.

Reply:
lol yeah not sure why I did that 👍